starkyModules/metadataRoles.ts

```diff
@@ -0,0 +1,83 @@
+import { retrieveAssets } from "../aspect/retrieveAssets";
+
+import { StarkyModuleConfig, StarkyModuleField } from "./types";
+
+export const name = "Metedata Roles";
+
+export const fields: StarkyModuleField[] = [
+  {
+    id: "contractAddress",
+    question: "What's the ERC-721 contract address?",
+  },
+  {
+    id: "conditionPattern",
+    question: "What's the path & pattern to get the role?",
+    textarea: true,
+  },
+];
+
+export const configName = (
+  starknetNetwork: string,
+  starkyModuleConfig: StarkyModuleConfig
+): string => {
+  return `${starknetNetwork} - Metadata - ${starkyModuleConfig.contractAddress}`;
+};
+
+const tryParseJSONObject = (jsonString: string) => {
+  try {
+    var o = JSON.parse(jsonString);
+    if (o && typeof o === "object") {
+      return o;
+    }
+  } catch (e) {}
+  return false;
+};
+
+export const shouldHaveRole = async (
+  starknetWalletAddress: string,
+  starknetNetwork: "mainnet" | "goerli",
+  starkyModuleConfig: StarkyModuleConfig
+): Promise<boolean> => {
+  const ownedAssets = await retrieveAssets({
+    starknetNetwork,
+    contractAddress: starkyModuleConfig.contractAddress,
+    ownerAddress: starknetWalletAddress,
+  });
+
+  const conditions = tryParseJSONObject(starkyModuleConfig.conditionPattern);
+  if (!conditions) return false;
+
+  let isFulfilled = false;
+  ownedAssets.forEach((asset: any) => {
+    let isMatched = true;
+    for (const c of conditions) {
+      const pathElements = c.path.split(".");
+      let currentData = asset;
+
+      // traverse the asset object to get the value at the specified path
+      for (let i = 0; i < pathElements.length; i++) {
+        if (currentData.hasOwnProperty(pathElements[i])) {
+          currentData = currentData[pathElements[i]];
+        } else if (Array.isArray(currentData)) {
+          currentData = currentData.find(
+            (attribute) => attribute[pathElements[i]] === pathElements[i + 1]
+          ).value;
```

Comment:
not 100% sure I understand this one
If the path corresponds to an array, you loop on the children to find a children that has a key `pathElements[i]` (what happens if the child does not have such a key?) 

Maybe a comment for this part?

Reply:
It's based on the metadata standard OpenSea supports (I think Aspect and Mintsquare also follow it) : 
```
{
  "description": "description", 
  "external_url": "external_url", 
  "image": "image_uri", 
  "name": "name,
  "attributes": [ {"trait_type": "trait1", "value": "val1"}, {"trait_type": "trait2", "value": "val2"} ]
}
```
I added conditions to ensure children i and i + 1 exist and a comment but maybe there's a way to make it more general ?

Reply:
Ok got it, thanks I like it that way!